### ISSUE TITLE
Add type hints

### ISSUE THREAD

**Issue by kasium** [jaegertracing/jaeger-client-python#319]
## Requirement - what kind of business use case are you trying to solve?

As you might know many python3 only packages already have type hints present based on the typing module. Afterwards the types can be checked e.g. with mypy. It would be great if you could add type hints and mark the package as types (aka add a py.typed file)

## Problem - what in Jaeger blocks you from solving the requirement?

Missing type hints in the code base

## Proposal - what do you suggest to solve the problem or improve the existing situation?

Add type hints 😄

**Comment by kasium** [comment]
@yurishkuro I'm happy to give a contribution a try. However, I guess with the current support python versions it will get hard. I would recommend to drop at least python 3.5 support if not even 3.6. What do you think?

**Comment by yurishkuro** [comment]
yes, it's fine to raise min supported version (in separate PR)

**Comment by kasium** [comment]
So, should I just change the setup.py file or also other stuff?

**Comment by yurishkuro** [comment]
Also GitHub actions

**Comment by kasium** [comment]
@yurishkuro so I see the following needed types

- [X] reporter ( #333 )
- [ ] config ( #334 )
- [X] util ( #326 )
- [X] tracer ( #328 )
- [ ] span
- [ ] metircs
- [ ] throttler
- [ ] sampler

Can you please check if something is missing or is too much

**Comment by yurishkuro** [comment]
yep, looks just right

**Comment by kasium** [comment]
@yurishkuro I currently hit an issue with my latest types (local change)

```
ImportError while loading conftest '/root/git/jaeger-client-python/tests/conftest.py'.
tests/conftest.py:17: in <module>
    from jaeger_client import ConstSampler, Tracer
jaeger_client/__init__.py:18: in <module>
    from .tracer import Tracer  # noqa
jaeger_client/tracer.py:35: in <module>
    from .span import Span, SAMPLED_FLAG, DEBUG_FLAG
jaeger_client/span.py:26: in <module>
    from .tracer import Tracer
E   ImportError: cannot import name 'Tracer' from 'jaeger_client.tracer' (/root/git/jaeger-client-python/jaeger_client/tracer.py)
```

This happens because `tracer.py` imports `span.py`, but `span.py` needs `tracer.py` to type the constructor. So I would need the `if TYPE_CHECKING` construct which will decrease the coverage

**Comment by yurishkuro** [comment]
Does mypy have any means of adding types without explicit imports?

If no other solution, we'd have to take a hit on the coverage.

**Comment by yurishkuro** [comment]
I am curious why it works when `TYPE_CHECKING` is true, doesn't it still create a circular import?

> Does mypy have any means of adding types without explicit imports?

what I meant was if there is a way to specify types in quotes as fully qualified names, to bypass the need for import.

**Comment by kasium** [comment]
So, how mypy resolves import is whole chapter, but no, this doesn't result in circular imports. Putting names in quotes just helps to resolve reference issues or if imports should/can not be resolved during runtime but still imports are needed.

For examples, this will fail with an unknown reference
```python
def foo() -> Bar:
  ...
```

This will also fail
```python
if TYPE_CHECKING
  import Bar

def foo() -> Bar:
  ...
```

This will work
```python
if TYPE_CHECKING
  import Bar

def foo() -> 'Bar':
  ...
```

**Comment by yurishkuro** [comment]
> Proposal - what do you suggest to solve the problem or improve the existing situation?
> Add type hints 😄

I should've asked from the start - what is your exit criteria for this ticket? Was there some specific task you were trying to do that would benefit from types?

**Comment by kasium** [comment]
Valid and good question. I would say this ticket is done if all public API as discussed above is types, a `py.typed` is added (really minor task) and the whole stuff is released.

I used this project somewhere else and I would like to benefit from the types

**Comment by yurishkuro** [comment]
so looks like this is done?

**Comment by yurishkuro** [comment]
released as 4.7.0

**Comment by kasium** [comment]
Great, thanks a lot!
